Library/Homebrew/cmd/upgrade.rb

```diff
@@ -171,4 +181,100 @@ def upgrade_formula(f)
       nil
     end
   end
+
+  def relink_dependents(formulae)
+    return if formulae.empty?
+
+    kegs = Formula.installed
+                  .map(&:opt_or_installed_prefix_keg)
+                  .reject { |f| f.runtime_dependencies.to_a.empty? }
+
+    formulae_to_assess = formulae
+    wave = 0
+
+    until formulae_to_assess.empty?
+      wave += 1
```

Comment:
I'm not sure I understand why the `until`/`wave` is necessary, can you explain?

Reply:
The initial version performs the assessment in a depth-first fashion (libyaml->ruby->...). Imagine libyaml broke a dependent X, which can be upgraded, but ruby broke X too; it will be upgraded twice in that version (once for ruby and another for libyaml).

In this version, I check all formulae in that dependency level *in a single pass*, reinstall/upgrade them, and then loop again to check their dependents, until this results in no broken formulae. This is why I print all messages at once instead of in each loop, to give a complete context of what will be done in this step.

Finally, the wave stuff serves to mark which depth level we're in regarding the `brew upgrade`d formulae.

Reply:
Gotcha, I see. Whenever I see an `until`, though, I'm worried about the potential for an infinite loop. I think given you're storing the `wave` here you should `break` when `wave` hits a certain level?

Reply:
@amyspark I think the lowest level you think would still be useful. From your description that could be `2` or `3`? What do you think?



> Should we also add a --just-fix-them-all flag in case someone still wants to jump down this rabbit hole?

Let's pass on that unless we get a request for it.

Reply:
In my Mac the deepest level is 5, starting with `poppler`. I'll use that.

Reply:
@amyspark Can you explain; is that the number of levels of recursive dependencies?

Reply:
Yes. Suppose I upgrade `poppler` and it breaks a dependent, that's how deep I can go in the dependency tree before I run out of formulae to keep assessing and reinstalling/upgrading.

Reply:
Of course, this is assuming the very worst case!

Reply:
@amyspark Given this is going down/up a tree structure I think that a recursive solution might be more natural. What do you think?

Reply:
It is more natural, you're correct. However:
- we'll have to fetch all installed kegs *in each recursion*
- it's not easy to return the list of upgraded/reinstalled formulae
- if it were (perhaps a `[upgrade, reinstalled].flatten`), we have to deduplicate formulae so as to act only on the remaining ones.

Reply:
@amyspark I was thinking about this a bit more and I think the problem is putting `upgrade` and `reinstall` in the same loop. What's I'd propose:

- find all the outdated dependents (if `brew upgrade` is specified without arguments) or all the named formulae
- for each of those dependents, look at the formulae that depend directly on them and are outdated
- for each of those second-level formulae (dependents): recursively look at the formulae that depend on them, finding those that are outdated etc.
- when all formulae in the tree have been found (but not yet upgraded): upgrade them all
- when all formulae have been upgraded (or any upgrades failed): check the linkage on all formulae and reinstall those necessary

How does that sound?

Reply:
That sounds... so... sweeeet 😭 

This only leaves the messaging stuff. When a dependent is detected to be broken, how would you like it to be reported? Is the current version fine?

Reply:
@amyspark Current version is fine, yeh 👍 .



> when brew upgrade X Y Z, even if they are not X, Y or Z?

@amyspark Yes, I think so. I realise that's a behaviour change but it's one that avoids unnecessary complexity in the code while still respecting pins.